### PR TITLE
💜 Implementing the backend API to the frontend

### PR DESCRIPTION
**Pull request by Breadkenty**
This is still incomplete but I wanted to get this out here for @sleeping-dev to help me with this comment I made on discord:

Although I do need help with one thing, and I'm a little too afraid to mess up the backend too much but I think how we do the GET for `Character` and `User` should change since these two refers to their `displayNames` to call them rather than its `Id`, it makes doing API calls for them a little complicated when I get objects that are nested in something like a `Combo` where it includes a `CharacterId`. To make it more simpler, is it possible to do a GET on both the `Id` and the `DisplayName`? I was looking at the Pokemon API(https://pokeapi.co/docs/v2#pokemon-section) for my students earlier today and noticed it can be done in the endpoints like this(https://pokeapi.co/api/v2/ability/{id or name}/), but would like a little help with making it happen.  I can go ahead and make a pull request for my changes I made on the controller and the core to take the `DisplayName` as the parameter for the GET Users so we can build on that. Also @Sleep#3174 I think this'll make it easier for me but I'm gonna make you a collaborator on the app, that way you have some powers to merge also.

I've already made some changes to the `UserControllers` to take the `displayName` for GET rather than the `id` so that our URL parameters wouldn't be `id` based eg. `smashcombos.com/user/Breadkenty` instead of `smashcombos.com/user/3`

### CODE REVIEW

**Comment by skittlz444** [comment]
The main thing I have against using display name in the url is duplication. If display names are enforced to be unique then there's no issue, but we'd have to force them to be unique, and to only contain url friendly characters.

File: Api/ClientApp/src/pages/User.jsx [hunk 1]
```diff
@@ -35,10 +46,31 @@ export function User() {
...
+    fetch(`/api/Characters/${characterVariableName}`)
+      .then(response => response.json())
+      .then(apiData => {
+        console.log(apiData)
+      })
+  }
+
+  useEffect(getUser, [])
+
+  console.log(user)
```
Comment (skittlz444) [inline Api/ClientApp/src/pages/User.jsx:68]: Don't leave this log

File: Api/ClientApp/src/pages/User.jsx [hunk 2]
```diff
@@ -35,10 +46,31 @@ export function User() {
     event.preventDefault()
     console.log('Dismiss:')
```
Comment (skittlz444) [inline Api/ClientApp/src/pages/User.jsx:47]: Remove console logs

File: Api/ClientApp/src/pages/User.jsx [hunk 3]
```diff
@@ -35,10 +46,31 @@ export function User() {
...
+      .then(apiData => {
+        setUser(apiData)
+      })
+  }
+
+  function getCharacter(characterVariableName) {
+    fetch(`/api/Characters/${characterVariableName}`)
+      .then(response => response.json())
+      .then(apiData => {
+        console.log(apiData)
```
Comment (skittlz444) [inline Api/ClientApp/src/pages/User.jsx:62]: Remove console logs

File: Api/ClientApp/src/pages/User.jsx [hunk 4]
```diff
@@ -35,10 +46,31 @@ export function User() {
...
+  function getUser() {
+    fetch(`/api/Users/${displayName}`)
+      .then(response => response.json())
+      .then(apiData => {
+        setUser(apiData)
+      })
+  }
+
+  function getCharacter(characterVariableName) {
+    fetch(`/api/Characters/${characterVariableName}`)
```
Comment (skittlz444) [inline Api/ClientApp/src/pages/User.jsx:59]: I feel like this get characters function should be something exported from the characters component, this code shouldn't be duplicated into everywhere that needs to get a character.

File: Api/Controllers/UsersController.cs [hunk 5]
```diff
@@ -107,22 +107,27 @@ public async Task<ActionResult<User>> PostUser([FromBody] User user)
...
+        private bool UserExists(string displayName)
+        {
+            return _context.Users.Any(user => user.DisplayName == displayName);
+        }
+
+        private string GetCurrentUserDisplayName()
         {
             // Get the User Id from the claim and then parse it as an integer.
-            return int.Parse(User.Claims.FirstOrDefault(claim => claim.Type == "Id").Value);
+            return User.Claims.FirstOrDefault(claim => claim.Type == "DisplayName").Value;
```
Comment (skittlz444) [inline Api/Controllers/UsersController.cs:130]: Use single or default, so I can't set my name to someone else's and get their data.

File: Smash_Combos.Core/Cqrs/Users/GetUser/GetUserRequestHandler.cs [hunk 6]
```diff
@@ -30,7 +30,7 @@ public async Task<GetUserResponse> Handle(GetUserRequest request, CancellationTo
                         .Include(user => user.Comments)
                             .ThenInclude(comment => comment.Reports)
                         .Include(user => user.Infractions)
-                        .Where(user => user.Id == request.UserId)
+                        .Where(user => user.DisplayName == request.DisplayName)
                         .FirstOrDefaultAsync();
```
Comment (skittlz444) [inline Smash_Combos.Core/Cqrs/Users/GetUser/GetUserRequestHandler.cs:34]: Probably best to use single or default to get sensitive data like a user, we don't want to accidentally return the first if someone does something cheeky. Which is definately possible when using display name as the identifier.

**Review by skittlz444 — CHANGES_REQUESTED** [review]
Just a few changes to make the display name identification method more secure. Also you should make the column unique if it isn't already (don't have that up in front of me right now), this might mean removing some existing users that have doubled up already. Other than things for that it looks good!

**Comment by Breadkenty** [comment]
I added the GetReportByDisplayName controller and also adding it to the CQRS. I made this so that getting a list of reports fo a user isn't so difficult on the front end lol. 

@skittlz444 Thanks for the feedback. I'll apply them before I am finished with this PR.

**Comment by Breadkenty** [comment]
@sleeping-dev 

1. We’ve made a temporary fix to passing the character Object instead of its Id. This should be fixed. You've written a temporary fix for this by this code you wrote
```
 var user = await _dbContext.Users
                      .Include(user => user.Combos)
                          .ThenInclude(combo => combo.Reports)
                      .Include(user => user.Comments)
                          .ThenInclude(comment => comment.Reports)
                      .Include(user => user.Infractions)
                      .Where(user => user.DisplayName == request.DisplayName)
                      .FirstOrDefaultAsync();

            if (user == null)
                return null;



            var response = new GetUserResponse
            {
                Id = user.Id,
                DisplayName = user.DisplayName,
                UserType = user.UserType,
                Combos = new List<ComboDto>(),
                Comments = _mapper.Map<List<CommentDto>>(user.Comments),
                Infractions = _mapper.Map<List<InfractionDto>>(user.Infractions)
            };

            foreach (var combo in user.Combos)
            {
                var comboDto = _mapper.Map<ComboDto>(combo);
                var character = await _dbContext.Characters.Where(character => character.Id == combo.CharacterId).FirstOrDefaultAsync();
                comboDto.Character = _mapper.Map<CharacterDto>(character);
                response.Combos.Add(comboDto);
            }
            

            return response;
```

However this presented with another problem where the `Moderator` object is returning `null` so I changed it back temporarily. Id like to avoid getting Id's back for anything regarding `Users` (for `Moderator` and `Reporter`) and `Characters` so if you can help me with updating the backend would help me a ton here.

2. There's an object I'm missing that I would like to have when I `GET Report` which are the `Combo` and `Comment` associated with said `Report` This is because I'd like to link the content that the `Report` comes from. Refer to the `Reported Content` column in the screenshot.

Screen Shot 2020-09-03 at 1 26 31 AM

3. This issue might be related to the first one but some of the comments where certain comments are returning a null `User` object. It's looking like it's returning `User` Objects only if the `Combo.User.DisplayName` and the `Comment.User.DisplayName` matches
Screen Shot 2020-09-01 at 9 59 06 PM

4. I've already started to work on this, but ideally I'd like all `GET` requests for `Users`, "Infractions", "Reports", and `Characters` to be by the `DisplayName`/ `VariableName` instead of `Id`. 
- Get `User` by `DisplayName`
- Get `Character` by `VariableName`
- Get `Reports` by `DisplayName` (get all reports by a given user)
- Get `Infractions` by `DisplayName (get all infractions by a given user) 

These are all the things I can think of right now at least for GETs. I know I'll come across some stuff with POSTs and PUTs but we'll work on that when we get there. Any help with this would be great.

**Comment by sleeping-dev** [comment]
Alright @Breadkenty, bullet points 1-3 should be fixed/added. I added the Character object reference to Combos & fixed some missing .Include() statements so now everything should be loaded correctly.

I'll take a break now and see if I can fix the rest after dinner so you have everything you need frontend-wise.

**Comment by sleeping-dev** [comment]
I had an idea for handling the communication between the core and the API Controllers a bit better and implemented it for the Combo-Methods as a test. Would love a review on it @skittlz444, if it breaks the pattern or is faulty in other ways:

Basically I introduced a `ResponseBase` class with a Generic `Data` property to hold the response data, and a `ResponseStatus` and `ResponseMessage` property so we can deal with the Core-Responses in the API Controllers in a cleaner way. Before I was putting booleans in the responses, or checking for null, etc, this would be a nice streamlined way of going about things. What do you think?

**Comment by sleeping-dev** [comment]
This should take care of everything you requested Kento, let me know if anything's missing.

**Comment by Breadkenty** [comment]
@sleeping-dev Awesome job bud! I made a couple of commits to work off what you've done so far. As I was working on this, I was thinking of something that I want to run by you. 

Should we add a dismiss property for the Infraction model? I'm thinking this would be used in the following scenario:

1. A moderator wrongfully infracts a user
2. A user appealed their infraction by contacting the admins
3. The admins look at the user's infraction that was given by the moderator
4. Admins realize that the user was wrongfully infracted, therefore dismissing the infraction
5. The infraction will be on record just in case we come across another issue with the same exact user if they commit another offense
6. Address the issue with moderators to prevent abusive use of their power

For this reason, I went ahead and added the dismiss button on infractions. Let me know what you think. For the most part, I think everything in regards to GET is done with my recent commit. The next thing I'll work on is the ability to:
1. Create a report on comment/combo
2. Create an infraction
3. Dismiss infraction/report
4. Unban a user
5. Limiting the report UI to moderators and admins

My goal to get it done by Weds went right out the window but we'll get it done by this Sunday at the latest at this rate!

**Comment by Breadkenty** [comment]
@skittlz444 @skittlz444 Oh and one more thing. We had an issue with using `SingleOrDefault` specifically for the GET `User/DisplayName`. For the time being, I went ahead and switched it back to `FirstOrDefault` but if you can find the solution to it, that would be great. This was the error I was getting:
Screen Shot 2020-09-03 at 9 55 51 PM

**Comment by sleeping-dev** [comment]
I believe exception that's being thrown here is working as intended, could it be that you have multiple users with the same name in your database? It worked fine for me, but I just tried it while having two users with the same name and got the same exception. So this should be working as intended, we just need to surround it with a try-catch to catch the exception and react accordingly, I'll whip that up real quick.

**Comment by sleeping-dev** [comment]
@Breadkenty Yes, the comment thing is because when submitting the comment the email-field is missing:
```
  const [comment, setComment] = useState({
    comboId: parseInt(comboId),
    user: { // Email should be in here
      displayName: loggedInUser.displayName, 
    },
    body: '',
  })
```

This will fix itself when we move the Comment-Methods to the core and use CQRS, I don't have time to do it right now but I'll to id later today.

**Comment by Breadkenty** [comment]
@sleeping-dev You're completely right it was that I had multiple users with the same `DisplayName`, which is weird because I didn't create them and you shouldn't be able to create a user with an existing display name. Not sure how it got there. I'll have to look into this further later.

**Comment by sleeping-dev** [comment]
I didn't realize that you wanted everything to go by the DisplayName. If we commit to having the DisplayName be the main property to determine users by, we might as well make it the primary key in the database, that way we can ensure that there's no double users. Then we could even remove the Id...

**Review by Breadkenty — COMMENTED** [review]
@sleeping-dev For the sake of confusion we should probably rename these

File: Smash_Combos.Core/Cqrs/Combos/PutCombo/PutComboRequest.cs [hunk 7]
```diff
@@ -9,10 +9,14 @@ namespace Smash_Combos.Core.Cqrs.Combos.PutCombo
...
-        public int Id { get; set; }
-        public UserDto User { get; set; }
+        public int UserId { get; set; }
+
+        [Required]
+        public int ComboId { get; set; }
+
         [Required]
-        public Character Character { get; set; }
+        public string CharacterName { get; set; }
```
[the remark this answers is not in the document]
Comment (sleeping-dev) [reply Smash_Combos.Core/Cqrs/Combos/PutCombo/PutComboRequest.cs]: Done

File: Smash_Combos.Core/Cqrs/Combos/PutCombo/PutComboRequestHandler.cs [hunk 8]
```diff
@@ -25,23 +25,31 @@ public PutComboRequestHandler(IDbContext dbContext, IMapper mapper)
...
-                return new PutComboResponse { Data = null, ResponseStatus = ResponseStatus.NotFound, ResponseMessage = "Couldn't find Combo" };
+                return new PutComboResponse { ResponseStatus = ResponseStatus.NotFound, ResponseMessage = "Couldn't find Combo" };
 
-            if (combo.User.Id != user.Id && user.UserType != UserType.Admin)
-                return new PutComboResponse { Data = null, ResponseStatus = ResponseStatus.NotAuthorized, ResponseMessage = "Not authorized to edit this Combo" };
+            if (combo.User.Id != user.Id && user.UserType == UserType.User)
+                return new PutComboResponse { ResponseStatus = ResponseStatus.NotAuthorized, ResponseMessage = "Not authorized to edit this Combo" };
 
-            var character = await _dbContext.Characters.Where(character => character.VariableName == request.Character.VariableName).FirstOrDefaultAsync();
+            var character = await _dbContext.Characters.Where(character => character.VariableName == request.CharacterName).FirstOrDefaultAsync();
```
[the remark this answers is not in the document]
Comment (sleeping-dev) [reply Smash_Combos.Core/Cqrs/Combos/PutCombo/PutComboRequestHandler.cs]: Done

**Comment by Breadkenty** [comment]
Everything on the frontend should be up to date, all we're waiting for now is the error messages with the requests. 

Couple things I'd like to request changes in the backend:
1. Reports and Comments body shouldn’t be nullable
2. POST Sessions to return an error message if a user is banned. `User is currently banned for: [insert infraction body here]`
3. POST Infraction to dismiss all existing non-dismissed reports
4. Reports can only be made once by the user
5. Prevent mods/admins from warning/ban/infracting themselves. 
6. Prevent mods/admins from dismissing their own reports. (Hopefully, another mod will keep other mods in check)

@skittlz444 I'm so sorry that this PR is HUGE! I didn't think about how to split this into bite-size chunks, but I know for next time. We're almost at the finish line for this feature!

**Comment by skittlz444** [comment]
No need to apologise! You guys have been doing heaps of great work. I have to apologise though, with everything happening with me at the moment I really don't have enough time to review it all, I had a look at a few request handlers, controllers etc and after the problem detail pr is in here and a good full systems test is done to iron out the bugs then its probably good to go. The structure is there, I'll trust you guys for the details.

**Pull request merged**